core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs

```diff
@@ -0,0 +1,243 @@
+/// variant and is then mutated into `Rocksdb` once RocksDB cache is caught up. After which it
+/// can never revert back to `Postgres` as we assume RocksDB cannot fall behind under normal state
+/// keeper operation.
+#[derive(Debug, Clone)]
+pub enum AsyncRocksdbCache {
+    Postgres(ConnectionPool),
+    Rocksdb(RocksDB<StateKeeperColumnFamily>, ConnectionPool),
+}
+
+impl AsyncRocksdbCache {
+    /// Returns a [`ReadStorage`] implementation backed by Postgres
+    async fn access_storage_pg(
+        rt_handle: Handle,
+        pool: &ConnectionPool,
+    ) -> anyhow::Result<PgOrRocksdbStorage> {
+        let mut connection = pool.access_storage().await?;
+
+        // Check whether we performed snapshot recovery
+        let snapshot_recovery = connection
+            .snapshot_recovery_dal()
+            .get_applied_snapshot_status()
+            .await
+            .context("failed getting snapshot recovery info")?;
+        let (miniblock_number, l1_batch_number) = if let Some(snapshot_recovery) = snapshot_recovery
+        {
+            (
+                snapshot_recovery.miniblock_number,
+                snapshot_recovery.l1_batch_number,
+            )
+        } else {
+            let mut dal = connection.blocks_dal();
+            let l1_batch_number = dal.get_sealed_l1_batch_number().await?.unwrap_or_default();
+            let (_, miniblock_number) = dal
+                .get_miniblock_range_of_l1_batch(l1_batch_number)
+                .await?
+                .unwrap_or_default();
+            (miniblock_number, l1_batch_number)
+        };
+
+        tracing::debug!(%l1_batch_number, %miniblock_number, "Using Postgres-based storage");
+        Ok(
+            PostgresStorage::new_async(rt_handle, connection, miniblock_number, true)
+                .await?
+                .into(),
+        )
+    }
+
+    /// Catches up RocksDB synchronously (i.e. assumes the gap is small) and
+    /// returns a [`ReadStorage`] implementation backed by caught-up RocksDB.
+    async fn access_storage_rocksdb<'a>(
+        conn: &mut StorageProcessor<'_>,
+        rocksdb: RocksDB<StateKeeperColumnFamily>,
+        stop_receiver: &watch::Receiver<bool>,
+    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'a>>> {
+        tracing::debug!("Catching up RocksDB synchronously");
+        let rocksdb_builder = RocksdbStorage::builder(rocksdb)
+            .await
+            .context("Failed initializing RocksDB storage")?;
+        let rocksdb = rocksdb_builder
+            .synchronize(conn, stop_receiver)
+            .await
+            .context("Failed to catch up state keeper RocksDB storage to Postgres")?;
+        if let Some(rocksdb) = &rocksdb {
+            let rocksdb_l1_batch_number = rocksdb.l1_batch_number().await.unwrap_or_default();
+            tracing::debug!(%rocksdb_l1_batch_number, "Using RocksDB-based storage");
+        } else {
+            tracing::warn!("Interrupted")
```

Comment:
Nit: missing semicolon. Also, I wouldn't consider this log warranting WARN level; interruption is a part of the normal node function.

Also, completely subjectively, a let-else would read better here:

```rust
let Some(rocksdb) = rocksdb else {
    tracing::info!("Synchronizing RocksDB interrupted");
    return Ok(None);
};
// work with `rocksdb`
```